### ISSUE TITLE
url generation does not work from virtual sources

### ISSUE THREAD

**Issue by vindarel** [lektor/lektor#63]
I'm editing lektor.com's sources, adding alternatives. In `layout.html`:

```
<a href="{{'.'|url(alt='es')}}
```

but that leads to a cryptic error message on build:

   E blog/archives/index.html (AttributeError: 'NoneType' object has no attribute 'endswith')

When we go to `blog/archives/index.html` we get the stacktrace:

```
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/lektor/builder.py", line 700, in update
    yield ctx
  File "/usr/local/lib/python2.7/dist-packages/lektor/builder.py", line 922, in build_artifact
    build_func(artifact)
  File "/home/vince/bacasable/lektorwebsite-copy/packages/blog-archive/lektor_blog_archive.py", line 139, in build_artifact
    this=self.source)
  File "/usr/local/lib/python2.7/dist-packages/lektor/builder.py", line 590, in render_template_into
    this=this, **extra)
  File "/usr/local/lib/python2.7/dist-packages/lektor/environment.py", line 466, in render_template
    return self.jinja_env.get_or_select_template(name).render(ctx)
  File "/usr/local/lib/python2.7/dist-packages/jinja2/environment.py", line 969, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/local/lib/python2.7/dist-packages/jinja2/environment.py", line 742, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/vince/bacasable/lektorwebsite-copy/templates/blog-archive/index.html", line 1, in top-level template code
    {% extends "blog-layout.html" %}
  File "/home/vince/bacasable/lektorwebsite-copy/templates/blog-layout.html", line 10, in top-level template code
    {% set feed_url = create_atom_feed(
  File "/home/vince/bacasable/lektorwebsite-copy/templates/layout.html", line 48, in top-level template code
    {% block footer %}
  File "/home/vince/bacasable/lektorwebsite-copy/templates/layout.html", line 63, in block "footer"
    <a href="{{'.'|url(alt='fr')}}">Français</a> •
  File "/usr/local/lib/python2.7/dist-packages/lektor/environment.py", line 385, in <lambda>
    lambda ctx, *a, **kw: url_to(*a, **kw)),
  File "/usr/local/lib/python2.7/dist-packages/lektor/context.py", line 17, in url_to
    return ctx.url_to(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/lektor/context.py", line 149, in url_to
    rv = self.source.url_to(path, alt=alt, absolute=True)
  File "/usr/local/lib/python2.7/dist-packages/lektor/sourceobj.py", line 99, in url_to
    source = self.pad.get(posixpath.join(self.path, path), alt=alt)
  File "/usr/lib/python2.7/posixpath.py", line 77, in join
    elif path == '' or path.endswith('/'):
AttributeError: 'NoneType' object has no attribute 'endswith'
```

There's a pb with the blog plugin, and a cryptic message.

**Comment by mitsuhiko** [comment]
To copy paste what I wrote on IRC:

Basically '.'|url requires that this.path is set. Because the path is relative to the current page's canonical path. Virtual sources (like the blog archive) do not have a path so that feature cannot work there.

One way would be to give virtual sources paths.  For instance the blog archive virtual source could be resolved with a special prefix.  Eg: `pad.get('/blog/^archive')` or something like this.

**Comment by vindarel** [comment]
and the workaround you gave is to use `{% if this.path %}` like so:

```
      {% if this.path %}
      <a href="{{'.'|url(alt='fr')}}">Français</a> •
      {% endif %}
```

and voilà.

**Comment by mitsuhiko** [comment]
This is now implemented for Lektor 2.0. https://www.getlektor.com/docs/content/paths/

The plugin in the lektor website repo is updated to take advantage of this feature.
